template.tpl

```diff
@@ -403,56 +892,252 @@ const parseSchemaToMajorKeyValue = (schema) => {
   return schema;
 };
 
-const parseCustomEventAndEntities = (eventData, eventProperties, userProperties) => {
-  for (let prop in eventData) {
-    if (eventData.hasOwnProperty(prop)) {
-      if (data.includeSelfDescribingEvent && prop.indexOf('x-sp-self_describing_event_') === 0) {
-        eventProperties[cleanPropertyName(prop)] = eventData[prop];
+/*
+ * Returns whether a property name is a Snowplow self-describing event property.
+ */
+const isSpSelfDescProp = (prop) => {
+  return prop.indexOf('x-sp-self_describing_event_') === 0;
+};
+
+/*
+ * Returns whether a property name is a Snowplow context/entity property.
+ */
+const isSpContextsProp = (prop) => {
+  return prop.indexOf('x-sp-contexts_') === 0;
+};
+
+/*
+ * Given a list of entity references and an entity name,
+ * returns the index of a matching reference.
+ * Matching reference means whether the entity name starts with ref.
+ *
+ * @param entity {string} - the entity name to match
+ * @param refsList {Array} - an array of strings
+ */
+const getReferenceIdx = (entity, refsList) => {
+  for (let i = 0; i < refsList.length; i++) {
+    if (entity.indexOf(refsList[i]) === 0) {
+      return i;
+    }
+  }
+  return -1;
+};
+
+/*
+ * Filters out invalid rules to avoid unintended behavior.
+ * (e.g. version control being ignored if version num is not included in name)
+ * Assumes that a rule contains 'key' and 'version' properties.
+ */
+const cleanRules = (rules) => {
+  return rules.filter((row) => {
+    if (row.version === 'control') {
+      // last char can't be null or empty string so fine for makeNumber
+      const lastAsNum = makeNumber(row.key.slice(-1));
+      if (lastAsNum != lastAsNum) {
```

Comment:
Thank you for reviewing @paulboocock !

This was meant to check for `NaN` (masked under a `null` which is not really null in gmt) without exluding zero. I replaced it now with `if (!lastCharAsNum && lastCharAsNum !== 0)`, to make it more explicit.
It aims to invalidate the case where a user specifies to `control` version without providing the version number in schema. Admittedly just naively checks whether last char is digit. Do you think this is ok?

Example valid rules:

 - `iglu:com.youtube/youtube/jsonschema/1-0-0` with option to apply version `control`
 - `contexts_com_youtube_youtube_1` with option to apply version `control`
 - `iglu:com.youtube/youtube/jsonschema/1-0-0` with option to apply independent of version
 - `contexts_com_youtube_youtube_1` with option to apply independent of version
 - `contexts_com_youtube_youtube` with option to apply independent of version

Invalid rules examples:

 - `contexts_com_youtube_youtube` with option to apply version `control`
 - `iglu:com.youtube/youtube/jsonschema/` with option to apply version `control`
 - `iglu:com.youtube/youtube/jsonschema/` with option to apply independent of version

Reply:
This is fine. All those invalid examples look very invalid to me. The only valid one which looks a bit dodgy is: `contexts_com_youtube_youtube` but I can see that it might be useful! Happy to leave it like this as long as its well documented.